src/excmds.ts

```diff
@@ -961,6 +974,185 @@ export function jumpprev(n = 1) {
+        scrollX: window.scrollX,
+        scrollY: window.scrollY,
+        tabId: currentTabId
+    };
+    try {
+        const tab = await browserBg.tabs.get(mark.tabId);
+        return onTabExists(tab);
+    } catch (e) {
+        return onTabNoLongerValid();
+    }
+
+    async function onTabExists(tab) {
+        const tabUrl = tab.url.split("#")[0]
+        if (mark.url !== tabUrl) {
+            return onTabNoLongerValid();
+        }
+        return goToTab(tab.id).then(() => {
+            scrolltab(tab.id, mark.scrollX, mark.scrollY, `Jumped to mark '${key}'`);
+        });
+    }
+
+    // the tab with mark's tabId doesn't exist or it has a different url than the mark's url
+    async function onTabNoLongerValid() {
+        const matchingTabs = await browserBg.tabs.query({url: mark.url});
+        // If there are no matching tabs, open a new one and update the mark's tabId for future use in this session
+        if (!matchingTabs.length) {
+            return openInNewTab(mark.url).then(updateMarkAndScroll());
+        }
+        // If there are multiple tabs open with the same url, just pick the first one and update the mark's tabId
+        // for future use in this session
+        return goToTab(matchingTabs[0].id).then(updateMarkAndScroll());
+    }
+
+    function updateMarkAndScroll() {
+        return (tab) => {
+            mark.tabId = tab.id;
+            state.globalMarks = globalMarks;
+            scrolltab(tab.id, mark.scrollX, mark.scrollY, `Jumped to mark '${key}'`);
+        };
+    }
+}
+
+/**
+ * Jumps to a location saved before the last mark jump as long as the tab it's located in exists and its url didn't change.
+ * Overwrites the location before the last mark jump - repeating this method will jump back and forth between two locations.
+ */
+//#content
+async function beforemarkjump() {
+    const beforeJumpMark = await State.getAsync("beforeJumpMark");
+    if (!beforeJumpMark) {
+        return;
+    }
+    try {
+        const tab = await browserBg.tabs.get(beforeJumpMark.tabId);
+        const tabUrl = tab.url.split("#")[0]
+        const {url, scrollX, scrollY, tabId} = beforeJumpMark;
+        if (url !== tabUrl) {
+            return;
+        }
+        const currentTabId = await activeTabId();
+        state.beforeJumpMark = {url: window.location.href.split("#")[0], scrollX: window.scrollX, scrollY: window.scrollY, tabId: currentTabId};
+        goToTab(tabId).then(() => scrolltab(tabId, scrollX, scrollY, "Jumped to the last location before a mark jump"));
+    } catch (e) {
+        // the mark's tab is no longer valid
+    }
+}
+
+/**
+ * Scrolls to a given position in a given tab and prints a message in it.
+ */
+//#content
+export async function scrolltab(tabId: number, scrollX: number, scrollY: number, message: string) {
+    browserBg.tabs.executeScript({
+        code: `window.scrollTo(${scrollX},${scrollY})`
+    }).then(() => acceptExCmd(tabId, [`fillcmdline_tmp 3000 ${message}`]));
```

Comment:
Is this where scrolling within acceptExCmd meant that the message appeared in the wrong tab? I'll try to look into it

Reply:
@mariuszdb I think I still need an answer to this - I'd prefer to avoid executeScript as we don't use it anywhere else. Thanks!